src/include/buffer.h

```diff
@@ -1133,31 +1125,23 @@ inline namespace v14_2_0 {
     operator seastar::net::packet() &&;
 #endif
 
-    iterator begin() {
-      return iterator(this, 0);
+    iterator begin(size_t offset=0) {
```

Comment:
Well, the `offset` introduction to `buffer::list` follows the established convention we already have in `buffer::ptr::begin()`.
The idea with extending constructor looks nice but we would need to rework iterator of `buffer::ptr` accordingly, I guess.
Alternative is to change return type of `advance()` (it returns `void` now) to allow chaining:

```cpp
bl.begin().advance(offset).copy(...)
```

Reply:
> The idea with extending constructor looks nice but we would need to rework iterator of buffer::ptr accordingly, I guess.

i don't think we need to "rework iterator of buffer::ptr accordingly". because
```c++
iterator_impl(bl_t *l, unsigned o=0)
```
it's already ready.

Reply:
So the situation is symmetrical – `buffer::ptr` already offers both conventions. I don't have strong feelings about neither of these two styles. `bl.cbegin(offset)` just looks shorter to type than `ceph::bufferlist::const_iterator(bl, offset).copy(...)`. Still, I would be fine with switching to the iterator style.

Reply:
the reason i don't like it is that

-  `cbegin()` and `begin()` are standard-compliant interfaces. i want to adhere to the least-surprise principle.
- we are adding interfaces which replicate existing ones from the functionality perspective. and more importantly, they are public interfaces.

yes, we are extending these interfaces instead of changing them. but, for me, it's just like `vector::size(ssize_t discount)` . yeah, i admit that `buffer::ptr()` already offers these extensions. that does not prove that it's an ideal design.

Reply:
> i want to adhere to the least-surprise principle

Me too. Personally, keeping in mind the picture of `ptr::begin()`, I got surprised by lack of the parameter on `list::begin()`. The principle can be seen from at least two different angles:
* coherence with the standard library,
* coherence in the `buffer` library.

I think we currently have a divergence: bringing the `offset` extension improves coherency in the boundaries of the `buffer` library while hurts in the `std::` aspect.

> already offers these extensions. that does not prove that it's an ideal design.

I agree it's far away from good design. The discount illustrates this issue nicely.

Maybe there are few users of the `offset`-in-`bptr::begin` nastiness? GCC is sniffing for them. Stay tuned! :-)

```
-    iterator begin(size_t offset=0) {
-      return iterator(this, offset, false);
+    iterator begin() {
+      return iterator(this, 0, false);
     }
-    const_iterator begin(size_t offset=0) const {
-      return const_iterator(this, offset, false);
+    const_iterator begin() const {
+      return const_iterator(this, 0, false);
     }
     const_iterator cbegin() const {
       return begin();
     }
-    const_iterator begin_deep(size_t offset=0) const {
+    const_iterator begin_deep() const {
+      return const_iterator(this, 0, true);
+    }
+
+    // to deprecate
+    iterator begin(size_t offset) __attribute__((deprecated("BL HUNT: offset"))) {
+      return iterator(this, offset, false);
+    }
+    const_iterator begin(size_t offset) const __attribute__((deprecated("BL HUNT: offset"))) {
+      return const_iterator(this, offset, false);
+    }
+    const_iterator begin_deep(size_t offset) const __attribute__((deprecated("BL HUNT: offset"))) {
       return const_iterator(this, offset, true);
     }
```

Reply:
@tchaikov: huh, it looks the `offset` extension is dead as dodo – nobody uses it at the moment!

```
$ grep "BL HUNT: offset" /tmp/20200131_bl_multihunt.txt |  cut -d: -f 1,2 | sort | uniq
$
```